### PR TITLE
[Rgen] Provide a way to parse the generic BindingTypeAttribute and an union struct.

### PR DESCRIPTION
**Pull request by mandel-macaque**
We want to be able to store as parte of the code change struct any information related to the binding type. To do so we have a new BindingTypeAttribute that is generic to allow extra config.

We also provide a new struct that works as a union of all the supported generic BindingTypeAttribute, that way we can add that new struct (in a follow up PR) to the code change struct.

### CODE REVIEW

File: src/rgen/Microsoft.Macios.Generator/DataModel/BindingData.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	public static implicit operator BindingTypeData<ObjCBindings.Protocol> (BindingData data) => data.protocolData;
+	public static implicit operator BindingTypeData<ObjCBindings.Category> (BindingData data) => data.categoryData;
+
+	public static implicit operator BindingData (BindingTypeData data) => new (BindingType.Unknown, data);
+	public static implicit operator BindingData (BindingTypeData<ObjCBindings.Class> data) => new (data);
+	public static implicit operator BindingData (BindingTypeData<ObjCBindings.Protocol> data) => new (data);
+	public static implicit operator BindingData (BindingTypeData<ObjCBindings.Category> data) => new (data);
+
+	/// <inheritdoc />
+	public bool Equals (BindingData other)
```
[the remark this answers is not in the document]
Comment (mandel-macaque) [reply src/rgen/Microsoft.Macios.Generator/DataModel/BindingData.cs:56]: Its a structure  ;)

**Review by rolfbjarne — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Comment by mandel-macaque** [comment]
Failing test is known: https://github.com/xamarin/xamarin-macios/pull/21870

**Pull request merged**